Relative links to `./[2]abc.md` crash the app

Hey :wave:,

I have the following directories & files:

![image](https://user-images.githubusercontent.com/788827/155911593-08a831e0-f03a-4902-a8d5-92d3937490c8.png)

In `services/docs/src/[1]introduction/[1]what-is-webstone-education.md`, I have a relative link like so:

```markdown
The best part? You get to [choose your own technology stack](./[3]build-your-own-curriculum.md) :tada:!
```

When I load Vitebook, it crashes with the following error:

```
services/docs dev: node:fs:585
services/docs dev:   handleErrorFromBinding(ctx);
services/docs dev:   ^
services/docs dev: Error: ENOENT: no such file or directory, open '/workspace/platform/services/docs/src/[1]introduction/%5B3%5Dbuild-your-own-curriculum.md'
services/docs dev:     at Object.openSync (node:fs:585:3)
services/docs dev:     at Object.readFileSync (node:fs:453:35)
services/docs dev:     at readRecentlyChangedFile (file:///workspace/platform/node_modules/.pnpm/@vitebook+core@0.23.2/node_modules/@vitebook/core/dist/node/index.js:8294:22)
services/docs dev:     at filePathToRoute (file:///workspace/platform/node_modules/.pnpm/@vitebook+core@0.23.2/node_modules/@vitebook/core/dist/node/index.js:8331:30)
services/docs dev:     at Object.parser.renderer.rules.link_open (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:17063:23)
services/docs dev:     at Renderer.renderInline (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:1000:32)
services/docs dev:     at Renderer.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:1025:26)
services/docs dev:     at Renderer.parser.renderer.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:17277:25)
services/docs dev:     at MarkdownIt2.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:4746:28)
services/docs dev:     at MarkdownIt2.parser.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:16914:20) {
services/docs dev:   errno: -2,
services/docs dev:   syscall: 'open',
services/docs dev:   code: 'ENOENT',
services/docs dev:   path: '/workspace/platform/services/docs/src/[1]introduction/%5B2%5Dbuild-your-own-curriculum.md'
services/docs dev: }
services/docs dev: Failed
/workspace/platform/services/docs:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @webstone/docs@0.0.0 dev: `vitebook dev -p 8080`
Exit status 1
 ELIFECYCLE  Command failed with exit code 1.
gitpod /workspace/platform $ 
```

If I remove the `.md` ending, the page loads but the link is still broken. I'd like to keep the `.md` extension so that links also work when someone views the file in GitHub.

I thought I'd report this here before I start to investigate and will follow up with any findings I may have.

--- a/packages/core/src/node/app/create/resolvePages.ts
+++ b/packages/core/src/node/app/create/resolvePages.ts
@@ -76,9 +76,10 @@ async function resolvePage(app: App, filePath: string): Promise<void> {
 
     const id = ensureLeadingSlash(path.relative(app.dirs.root.path, filePath));
     const relativeFilePath = app.dirs.root.relative(filePath);
+    const fileContent = (await readRecentlyChangedFile(filePath)).toString();
     const route =
       app.options.resolveRoute?.({ filePath, relativeFilePath }) ??
-      (await filePathToRoute(app, filePath, true));
+      (await filePathToRoute(app, filePath, fileContent, true));
 
     const page = await plugin.resolvePage?.({
       id,
@@ -146,9 +147,9 @@ export function stripOrderFromRoute(route: string) {
 export async function filePathToRoute(
   app: App,
   filePath: string,
+  fileContent: string,
   keepOrder = false,
 ): Promise<string> {
-  const fileContent = (await readRecentlyChangedFile(filePath)).toString();
   const fileExt = path.extname(filePath);
 
   let configuredRoute;

--- a/packages/markdown/src/node/parser/plugins/linksPlugin.ts
+++ b/packages/markdown/src/node/parser/plugins/linksPlugin.ts
@@ -1,5 +1,6 @@
 import { filePathToRoute, isLinkExternal } from '@vitebook/core/node';
 import { resolveRelativePath } from '@vitebook/core/node/utils';
+import fs from 'fs';
 import type { PluginWithOptions } from 'markdown-it';
 
 import type { MarkdownParserEnv } from '../types';
@@ -63,7 +64,7 @@ export const linksPlugin: PluginWithOptions<LinksPluginOptions> = (
       const hrefAttr = token.attrs![hrefIndex];
       const hrefLink = hrefAttr[1];
 
-      const internalLinkMatch = hrefLink.match(
+      const internalLinkMatch = decodeURIComponent(hrefLink).match(
         /^((?:.*)(?:\/|\.md|\.html))(#.*)?$/,
       );
 
@@ -97,7 +98,8 @@ export const linksPlugin: PluginWithOptions<LinksPluginOptions> = (
           ? app!.dirs.src.resolve('.' + rawPath)
           : resolveRelativePath(filePath!, rawPath);
 
-        const route = filePathToRoute(app!, absolutePath);
+        const fileContent = fs.readFileSync(absolutePath).toString();
+        const route = filePathToRoute(app!, absolutePath, fileContent);
 
         withRouterLink();
 

